File: clj/src/day02/level_reports_test.clj
(ns day02.level-reports-test
  (:require [clojure.test :refer [deftest is testing]]
            [day02.level-reports :as lr]))

(deftest day02.level-reports
  (testing "report-safe?
    a report only counts as safe if both of the following are true:
    - The levels are either all increasing or all decreasing.
    - Any two adjacent levels differ by at least one and at most three.
    "
    (is (= true
           (lr/report-safe? {:level-report [7 6 4 2 1]}))
        "levels are all decreasing")
    (is (= true
           (lr/report-safe? {:level-report [1 3 6 7 9]}))
        "levels are all increasing")
    (is (= false
           (lr/report-safe? {:level-report [1 2 7 8 9]}))
        "2 7 is an increase of 5")
    (is (= false
           (lr/report-safe? {:level-report [9 7 6 2 1]}))
        "not uniformly decreasing")
    (testing "The Problem Dampener is a reactor-mounted module that 
      lets the reactor safety systems tolerate a single bad level in what would otherwise be a safe report.
      "
      (is (= true
             (lr/report-safe? {:level-report [1 3 2 4 5]}))
          "not uniformly increasing, safe by removing the second level, 3")
      (is (= true
             (lr/report-safe? {:level-report [8 6 4 4 1]}))
          "4 4 is neither a decrease or increase, safe by removing the third level, 4")))
  (testing "safe-reports-count"
    (let [level-reports (lr/parse-level-reports-file {:file-name "src/day02/test.txt"})]
      (is (= {:level-reports [[7 6 4 2 1]
                              [1 2 7 8 9]
                              [9 7 6 2 1]
                              [1 3 2 4 5]
                              [8 6 4 4 1]
                              [1 3 6 7 9]]
              :safe-reports-count {:safe-reports 4
                                   :unsafe-reports 2}}
             {:level-reports level-reports
              :safe-reports-count (lr/safe-reports-count {:level-reports level-reports})})))))

File: js/day02/levelReports.test.js
require('../init');
const {
  parseLevelReportsFile,
  countSafeReports,
  reportIsSafe,
} = require('./levelReports');

describe('day02/levelReports', () => {
  describe('reportIsSafe', () => {
    it(`a report only counts as safe if both of the following are true:
    - The levels are either all increasing or all decreasing.
    - Any two adjacent levels differ by at least one and at most three.`, () => {
      expect(reportIsSafe({ levelReport: [7, 6, 4, 2, 1] })).to.eql(
        true,
        'levels are all decreasing'
      );
      expect(reportIsSafe({ levelReport: [1, 3, 6, 7, 9] })).to.eql(
        true,
        'levels are all increasing'
      );
      expect(reportIsSafe({ levelReport: [1, 2, 7, 8, 9] })).to.eql(
        false,
        `2 7 is an increase of 5`
      );
      expect(reportIsSafe({ levelReport: [9, 7, 6, 2, 1] })).to.eql(
        false,
        `6 2 is a decrease of 4`
      );
    });

    it(`The Problem Dampener is a reactor-mounted module that 
      lets the reactor safety systems tolerate a single bad level in what would otherwise be a safe report`, () => {
      expect(reportIsSafe({ levelReport: [1, 2, 7, 8, 9] })).to.eql(
        false,
        `2 7 is an increase of 5, unsafe regardless of which level is removed.`
      );
      expect(reportIsSafe({ levelReport: [9, 7, 6, 2, 1] })).to.eql(
        false,
        `6 2 is a decrease of 4, unsafe regardless of which level is removed.`
      );
      expect(reportIsSafe({ levelReport: [1, 3, 2, 4, 5] })).to.eql(
        true,
        `safe by removing the second level, 3`
      );
      expect(reportIsSafe({ levelReport: [8, 6, 4, 4, 1] })).to.eql(
        true,
        `safe by removing the third level, 4`
      );
    });
  });

  describe('countSafeReports', () => {
    it(`a report only counts as safe if both of the following are true:
    - The levels are either all increasing or all decreasing.
    - Any two adjacent levels differ by at least one and at most three.`, () => {
      const { levelReports } = parseLevelReportsFile({
        fileName: 'data/day02/test.txt',
      });
      expect({ levelReports }).to.eql({
        levelReports: [
          [7, 6, 4, 2, 1],
          [1, 2, 7, 8, 9],
          [9, 7, 6, 2, 1],
          [1, 3, 2, 4, 5],
          [8, 6, 4, 4, 1],
          [1, 3, 6, 7, 9],
        ],
      });
      const safeReports = countSafeReports({ levelReports });
      expect(safeReports).to.eql({
        safeReports: 4,
        unsafeReports: 2,
      });
    });
  });
});
